### PR TITLE
#76: Externally removed resources cause crash 

### PR DESCRIPTION
**Pull request by mdrlzy**
#76

### CODE REVIEW

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt [hunk 1]
```diff
@@ -136,14 +142,26 @@ class ResourcesPresenter(
...
             router.exit()
     }
 
-    private fun onSelectionChange(selection: Set<ResourceId>) {
+    suspend fun onRemovedResourceDetected() {
+        viewState.setProgressVisibility(true, "Indexing")
+
+        index.reindex()
+        // update current tags storage
+        tagsStorageRepo.provide(rootAndFav)
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt:155]: Am I right this line triggers garbage collection in tag storage?

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/GalleryPresenter.kt [hunk 2]
```diff
@@ -150,7 +164,39 @@ class GalleryPresenter(
         viewState.displayPreviewTags(resource.id, tags)
     }
 
-    private fun onPreviewsItemClick(itemView: PreviewItemView) {
+    private fun onRemovedResourceDetected() = presenterScope.launch {
+        viewState.setProgressVisibility(true, "Indexing")
+
+        index.reindex()
+        // update current tags storage
+        tagsStorageRepo.provide(rootAndFav)
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/presenter/GalleryPresenter.kt:172]: Am I right this line triggers garbage collection in tag storage?

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    override fun getNewListSize(): Int = newList.size
+
+    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int):
+        Boolean {
+        return oldList[oldItemPosition] == newList[newItemPosition]
+    }
+
+    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int):
+        Boolean {
+        return true
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt:21]: Should it be so?

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    private val oldList: List<ResourceMeta>,
+    private val newList: List<ResourceMeta>
+) : DiffUtil.Callback() {
+    override fun getOldListSize(): Int = oldList.size
+
+    override fun getNewListSize(): Int = newList.size
+
+    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int):
+        Boolean {
+        return oldList[oldItemPosition] == newList[newItemPosition]
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt:16]: Why comparing by reference and not by value?

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt (hunk 1 above)
Comment (mdrlzy) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/ResourcesPresenter.kt:155]: Yes, also registers new resources. I think we need a clearer way to do this. Perhaps the tags storage should have a reference to the index.

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/GalleryPresenter.kt (hunk 2 above)
Comment (mdrlzy) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/GalleryPresenter.kt:172]: Yes, also registers new resources. I think we need a clearer way to do this. Perhaps the tags storage should have a reference to the index.

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt (hunk 4 above)
Comment (mdrlzy) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt:16]: In Kotlin `==` compares by calling `equals` under the hood. To compare by reference you need to use `===`

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/tags/PlainTagsStorage.kt [hunk 5]
```diff
@@ -59,10 +59,9 @@ class PlainTagsStorage(
         val knownResources = tagsById.keys.toSet()
 
         val lostResources = knownResources.subtract(indexedResources)
-        if (lostResources.isNotEmpty()) {
+        if (lostResources.isNotEmpty())
             Log.d(TAGS_STORAGE, "lostResources: $lostResources")
-            throw AssertionError("Index lost resources")
-        }
+        lostResources.forEach { resource -> tagsById.remove(resource) }
```
[the remark this answers is not in the document]
Comment (mdrlzy) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/tags/PlainTagsStorage.kt:64]: I think the user does not need to know about it. Perhaps we need to create a global tags storage for lost resources. And when indexing new resources, check there.

**Comment by cnzqa** [comment]
@kirillt this is ok.

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt (hunk 4 above)
Comment (kirillt) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/adapter/ResourceMetaDiffUtilCallback.kt]: Oh, indeed :)

File: app/src/main/java/space/taran/arknavigator/mvp/presenter/GalleryPresenter.kt (hunk 2 above)
Comment (kirillt) [reply app/src/main/java/space/taran/arknavigator/mvp/presenter/GalleryPresenter.kt:172]: I think index and storage will be merged into one bigger entity ("database"?) in upcoming library.

**Review by kirillt — APPROVED** [review]
Good job

**Pull request merged**